File: 3427-special-array-ii/3427-special-array-ii.java
class Solution {
    public boolean[] isArraySpecial(int[] nums, int[][] queries) {
        int n = queries.length;
        boolean[] ans = new boolean[n];
        int[] converted = new int[nums.length];
        converted[0] = 0;

        int j = 0;
        for (int i = 1; i < nums.length; i++) {
            if ((nums[i] % 2) == (nums[i - 1] % 2)) {
                j++;
            }
            converted[i] = j;
        }

        for (int i = 0; i < n; i++) {
            ans[i] = (converted[queries[i][0]] == converted[queries[i][1]]) ? true : false;
        }

        return ans;
    }
}

File: 3427-special-array-ii/3427-special-array-ii.cpp
class Solution {
public:
    vector<bool> isArraySpecial(vector<int>& nums, vector<vector<int>>& queries) {
        int n = nums.size();
        int m = queries.size();
        if(n == 1)
        {
            vector<bool> ans(m, true);
            return ans;
        }
        vector<bool> ans(m);
        vector<int> temp(n);
        temp[0] = 0;
        int value = 0;
        for(int i=1;i<n;i++)
        {
            value += (nums[i-1] % 2 == nums[i] % 2);
            temp[i] = value;
        }
        for(auto num : temp) cout << num << " ";
        cout << endl;
        // 1 for diff same, 0 for diff
        int i = 0;
        for(auto query : queries)
        {
            int first =query[0];
            int second = query[1];
            cout << first << " " << second << endl;
            ans[i] = (temp[first] == temp[second]) ? true : false;
            i++;
        }
        return ans;
    }
};